### PR TITLE
Ports and Connections Part 2

### LINKED ISSUE #361: Port and connections. Part 2.
### Summary & Value
- Ability to drag a port to create an edge.
- When zooming out the width of the edges and the size of the arrows should grow slightly (talk with wdanilo about details)
- Both arrows of the connections as well as the connections should have a wide invisible area allowing for clicking with mouse. 
- When hovering over the ports / edges they should glow. In particular when hovering over the first / second half of the edge - it should glow.

### Specification
- The ports should be floating around nodes.
- The whole mouse integration should use the FRP system.

### Acceptance Criteria & Test Cases
- Ability to hover the ports and drag them in the GUI.

### PR DESCRIPTION
**Pull request by MichaelMauderer**
### Pull Request Description
This PR adds the ability to render connections between ports, adds UI for creating connections between ports, adds glow effects for ports and connections.
Closes #361 

### Important Notes
This is a draft. 

### Checklist
Please include the following checklist in your PR:

- [ ] The documentation has been updated if necessary.
- [ ] All code conforms to the [Rust](https://github.com/luna/enso/blob/master/doc/rust-style-guide.md), [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [ ] All code has been tested where possible.

### CODE REVIEW

File: src/rust/ensogl/src/display/scene.rs [hunk 1]
```diff
@@ -42,6 +42,9 @@ use enso_frp as frp;
 
 pub trait MouseTarget : Debug + 'static {
     fn mouse_down(&self) -> Option<enso_frp::Source> { None }
+    fn mouse_over(&self) -> Option<enso_frp::Source> { None }
+    fn mouse_leave(&self) -> Option<enso_frp::Source> { None }
+    fn mouse_move_global(&self) -> Option<enso_frp::Source> { None }
```
Comment (wdanilo) [inline src/rust/ensogl/src/display/scene.rs:47]: What it is for? The operations here should be for mouse target only. Global mouse move should not be included here.

File: src/rust/ensogl/src/display/shape/primitive/def/var.rs [hunk 2]
```diff
@@ -473,6 +474,20 @@ where T: Acos<Output=T> {
...
+impl<T> Atan2 for Var<T>
+    where T: Atan2<Output=T> + RefInto<Glsl> + Copy{
+    type Output = Var<T>;
+    fn atan2(&self, other:Self) -> Self {
+        match (self, &other) {
+            (Var::Static(lhs), Var::Static(rhs)) => Var::Static(lhs.atan2(*rhs)),
+            _ => {
+                    let code = format!("atan({},{})",self.glsl(),other.glsl());
+                    Var::Dynamic(code.into())
+            }
```
Comment (wdanilo) [inline src/rust/ensogl/src/display/shape/primitive/def/var.rs:486]: make it single line pls (without `=>` alignment). Its not more clear the way its now.

File: src/rust/ensogl/src/gui/component.rs [hunk 3]
```diff
@@ -23,24 +23,35 @@ use enso_frp as frp;
...
-        Self {network,mouse_down}
+        Self {network,mouse_down,mouse_over,mouse_leave}
     }
 }
 
 impl MouseTarget for ShapeViewEvents {
     fn mouse_down(&self) -> Option<frp::Source> {
         Some(self.mouse_down.clone_ref())
     }
+    fn mouse_over(&self) -> Option<frp::Source> {
```
Comment (wdanilo) [inline src/rust/ensogl/src/gui/component.rs:49]: nl between fns

File: src/rust/ensogl/src/math/algebra.rs [hunk 4]
```diff
@@ -285,6 +285,28 @@ impl Acos for f32 {
     }
 }
 
+// ===============
+// === Atan2 ===
+// ===============
```
Comment (wdanilo) [inline src/rust/ensogl/src/math/algebra.rs:290]: alignment

File: src/rust/lib/graph-editor/src/component/node.rs [hunk 5]
```diff
@@ -133,9 +136,11 @@ pub mod shape {
 #[derive(Clone,CloneRef,Debug)]
 #[allow(missing_docs)]
 pub struct Events {
-    pub network    : frp::Network,
-    pub select     : frp::Source,
-    pub deselect   : frp::Source,
+    pub network      : frp::Network,
+    pub select       : frp::Source,
+    pub deselect     : frp::Source,
+    /// Emitted if a new port was created.
```
Comment (wdanilo) [inline src/rust/lib/graph-editor/src/component/node.rs:142]: this doc is not needed.

File: src/rust/lib/graph-editor/src/component/node/connection.rs [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+    ensogl::define_shape_system! {
+         (start:Vector2<f32>, end:Vector2<f32>, thickness:f32, glow:f32) {
+
+            let height     = Var::<f32>::from("input_start.y - input_end.y");
+            let width      = Var::<f32>::from("input_start.x - input_end.x");
+            let is_reverse = Var::<f32>::from("((((-sign(input_start.y-input_end.y)))))");
+
+            let angle_inner = Var::<f32>::from(90_f32.to_radians());
+            let triangle    = Triangle::<Var<f32>>::from_sides_and_angle(height,width,angle_inner);
+            let angle_a     = Var::<Angle<Radians>>::from(is_reverse * triangle.angle_a().clone());
```
Comment (wdanilo) [inline src/rust/lib/graph-editor/src/component/node/connection.rs:46]: why connection uses triangles and glows?

File: src/rust/lib/graph-editor/src/component/node/connection.rs [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+/// Connections are center aligned to make some internal computations easier.
+#[derive(Debug,Default,Clone,Copy)]
+pub struct ConnectionView {}
+impl ShapeViewDefinition for ConnectionView {
+    type Shape = shape::Shape;
+    fn new(_shape:&Self::Shape, _scene:&Scene,shape_registry:&ShapeRegistry) -> Self {
+        let shape_system = shape_registry.shape_system(PhantomData::<shape::Shape>);
+        shape_system.shape_system.set_alignment(
+            alignment::HorizontalAlignment::Center,alignment::VerticalAlignment::Center);
+
```
Comment (wdanilo) [inline src/rust/lib/graph-editor/src/component/node/connection.rs:108]: remove nl

File: src/rust/lib/graph-editor/src/component/node/connection.rs [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+/// Connections represent a link between an `InputPort` and an `OutputPort`.
+///
+/// Ports are shown as a connecting line going from one port to the other. The position snd shape
+/// of the `Connection` is defined by the start and end ports.
+///
+/// Connections can be created in the IDE by dragging from one port to another. This is currently handled in
+/// the `GraphEditor`, where events are collected and routed to allow dragging from ports,
+/// existence of a partially connected in-creation `Connection`, and finally the connecting of two
+/// ports.
```
Comment (wdanilo) [inline src/rust/lib/graph-editor/src/component/node/connection.rs:123]: There should be 2 ways to connect ports - either dragging, or by clicking (mouse down, mouse up) on one port, moving mouse (the connection follows mouse), and clicking on target port. Both should be implemented.

File: src/rust/lib/graph-editor/src/component/node/connection.rs [hunk 9]
```diff
@@ -0,0 +1,384 @@
...
+
+        let data         = ConnectionData { start_port,start,end_port,end,label,events,view,
+                                            logger };
+        let data         = Rc::new(data);
+        Self {data} . init() .init_frp()
+    }
+
+    fn init(self) -> Self {
+        self
+    }
```
Comment (wdanilo) [inline src/rust/lib/graph-editor/src/component/node/connection.rs:176]: a lot of strange things in this code :D I understand its draft, so Im not making any more comments about it.

File: src/rust/lib/graph-editor/src/component/node/connection.rs [hunk 10]
```diff
@@ -0,0 +1,384 @@
...
+    ///
+    /// Specifically:
+    /// * translates mouse events in hover events.
+    /// * sets up actions on port movement
+    /// * sets up actions on hover events
+    fn init_frp(self) -> Self {
+        let weak_connection = self.downgrade();
+        let network = &self.data.view.events.network;
+        frp::new_bridge_network! { [network,self.data.events.network]
+                let weak_connection_on_hover = weak_connection.clone();
```
Comment (wdanilo) [inline src/rust/lib/graph-editor/src/component/node/connection.rs:188]: the naming is starnge. Why cloned weak connection is a weak connection on hover?

File: src/rust/lib/graph-editor/src/component/node/connection.rs [hunk 11]
```diff
@@ -0,0 +1,384 @@
...
+    /// Set up the event handling for connections.
+    ///
+    /// Specifically:
+    /// * translates mouse events in hover events.
+    /// * sets up actions on port movement
+    /// * sets up actions on hover events
+    fn init_frp(self) -> Self {
+        let weak_connection = self.downgrade();
+        let network = &self.data.view.events.network;
+        frp::new_bridge_network! { [network,self.data.events.network]
```
Comment (wdanilo) [inline src/rust/lib/graph-editor/src/component/node/connection.rs:187]: Its late so Im may be missing something, but why do you need a bridge network here?

File: src/rust/lib/graph-editor/src/component/node/port.rs [hunk 12]
```diff
@@ -263,7 +295,7 @@ impl ShapeViewDefinition for OutputPortView {
 }
 
 /// Helper trait that describes a `ShapeViewDefinition` with a port shape.
-pub trait PortShapeViewDefinition = ShapeViewDefinition<Shape=shape::Shape>;
+pub trait PortShapeViewDefinition = ShapeViewDefinition<Shape=shape::Shape> + Clone;
```
Comment (wdanilo) [inline src/rust/lib/graph-editor/src/component/node/port.rs:298]: Why this? And why not CloneRef?

File: src/rust/lib/graph-editor/src/component/node/port.rs [hunk 13]
```diff
@@ -292,12 +324,52 @@ pub struct Port<T:PortShapeViewDefinition> {
...
 /// Type that represents an input port.
 pub type InputPort = Port<InputPortView>;
 
 /// Type that represents an output port.
 pub type OutputPort = Port<OutputPortView>;
 
+/// Enum over all possible port types.
+#[derive(Clone,CloneRef,Debug)]
+#[allow(missing_docs)]
+pub enum IOPort{
```
Comment (wdanilo) [inline src/rust/lib/graph-editor/src/component/node/port.rs:356]: This is not proper name in rust terms. I hate it too, but it should be `IoPort`. Maybe is there a better name?

File: src/rust/ensogl/src/display/scene.rs (hunk 1 above)
Comment (MichaelMauderer) [reply src/rust/ensogl/src/display/scene.rs:47]: Was a half baked attempt at getting mouse events directly to shapes. Has been removed.

File: src/rust/lib/graph-editor/src/component/node/connection.rs (hunk 6 above)
Comment (MichaelMauderer) [reply src/rust/lib/graph-editor/src/component/node/connection.rs]: No more triangles.
Glow is there as a hover effect.

File: src/rust/lib/graph-editor/src/component/node/connection.rs (hunk 11 above)
Comment (MichaelMauderer) [reply src/rust/lib/graph-editor/src/component/node/connection.rs:179]: Actually I think I don't. Has been changed to `extend`.

File: src/rust/lib/graph-editor/src/component/node/connection.rs (hunk 10 above)
Comment (MichaelMauderer) [reply src/rust/lib/graph-editor/src/component/node/connection.rs:180]: This has been cleaned up.

File: src/rust/lib/graph-editor/src/component/node/port.rs (hunk 13 above)
Comment (MichaelMauderer) [reply src/rust/lib/graph-editor/src/component/node/port.rs:373]: Maybe this should even be the main `Port` struct. Need to consider the naming and implications.

File: src/rust/lib/graph-editor/src/component/node/port.rs (hunk 12 above)
Comment (MichaelMauderer) [reply src/rust/lib/graph-editor/src/component/node/port.rs]: The additional bound is required for `WeakRef` etc. But `CloneRef` is enough.
Has been changd.

**Comment by wdanilo** [comment]
Im. going to close it. I will be working on that on other branch

**Pull request closed without merging**